server/core/src/main/java/dev/slimevr/tracking/trackers/hid/HIDCommon.kt

```diff
@@ -0,0 +1,288 @@
+import io.github.axisangles.ktmath.Vector3
+import java.nio.ByteBuffer
+import java.util.function.Consumer
+import kotlin.collections.set
+import kotlin.math.PI
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlin.math.sqrt
+
+class HIDCommon {
+	companion object {
+		const val HID_TRACKER_RECEIVER_VID = 0x1209
+		const val HID_TRACKER_RECEIVER_PID = 0x7690
+
+		const val PACKET_SIZE = 16
+
+		private val AXES_OFFSET = fromRotationVector(-FastMath.HALF_PI, 0f, 0f)
+
+		private fun setUpSensor(
+			device: HIDDevice,
+			trackerId: Int,
+			sensorType: IMUType,
+			sensorStatus: TrackerStatus,
+			magStatus: MagnetometerStatus,
+			trackersConsumer: Consumer<Tracker>,
+		) {
+			// LogManager.info("[TrackerServer] Sensor $trackerId for ${device.name}, status $sensorStatus")
+			var imuTracker = device.getTracker(trackerId)
+			if (imuTracker == null) {
+				var formattedHWID = device.hardwareIdentifier.replace(":", "").takeLast(5)
+				imuTracker = Tracker(
+					device,
+					VRServer.getNextLocalTrackerId(),
+					device.name + "/" + trackerId,
+					"Tracker $formattedHWID",
+					null,
+					trackerNum = trackerId,
+					hasRotation = true,
+					hasAcceleration = true,
+					userEditable = true,
+					imuType = sensorType,
+					allowFiltering = true,
+					needsReset = true,
+					needsMounting = true,
+					usesTimeout = false,
+					magStatus = magStatus,
+				)
+				// usesTimeout false because HID trackers aren't "Disconnected" unless receiver is physically removed probably
+				// TODO: Could tracker maybe use "Timed out" status without marking as disconnecting?
+				// TODO: can be marked as "Disconnected" by timeout if the tracker has enabled activity timeouts
+				device.trackers[trackerId] = imuTracker
+				trackersConsumer.accept(imuTracker)
+				imuTracker.status = sensorStatus
+				LogManager
+					.info(
+						"[TrackerServer] Added sensor $trackerId for ${device.name}, type $sensorType",
+					)
+			}
+		}
+
+		fun processPacket(
+			dataReceived: ByteArray,
+			i: Int,
+			packetType: Int,
+			device: HIDDevice,
+			q: IntArray,
+			a: IntArray,
+			m: IntArray,
+			trackersConsumer: Consumer<Tracker>,
+		) {
+			val trackerId = 0 // no concept of extensions
+
+			// Register tracker
+			if (packetType == 0) { // Tracker register packet (device info)
+				val imu_id = dataReceived[i + 8].toUByte().toInt()
+				val mag_id = dataReceived[i + 9].toUByte().toInt()
+				val sensorType = IMUType.getById(imu_id.toUInt())
+				// only able to register magnetometer status, not magnetometer type
+				val magStatus = MagnetometerStatus.getById(mag_id.toUByte())
+				if (sensorType != null && magStatus != null) {
+					setUpSensor(device, trackerId, sensorType, TrackerStatus.OK, magStatus, trackersConsumer)
+				}
+			}
+
+			val tracker: Tracker? = device.getTracker(trackerId)
+			if (tracker == null) { // not registered yet
+				return
+			}
+
+			// Packet data
+			var batt: Int? = null
```

Comment:
I think it's time we use a struct/class for this... @sctanf ?

Reply:
i am not sure, since the data here can be partial/aggregate from multiple packets, and it is only used here to update the device or tracker struct below